drivers/rotary_encoder/rotary.c

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2016 Marc Poulhi�s
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_rotarty_encoder
+ * @{
+ *
+ * @file
+ * @brief       Low-level rotary encoder implementation
+ *
+ * @author      Marc Poulhi�s <dkm@kataplop.net>
+ *
+ * @}
+ */
+
+#include "rotary.h"
+
+/*
+ * Code mainly based on http://www.buxtronix.net/2011/10/rotary-encoders-done-properly.html
+ */
+
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define R_START 0x0
+
+#ifdef HALF_STEP
+// Use the half-step state table (emits a code at 00 and 11)
+#define R_CCW_BEGIN 0x1    
+#define R_CW_BEGIN 0x2     
+#define R_START_M 0x3      
+#define R_CW_BEGIN_M 0x4    
+#define R_CCW_BEGIN_M 0x5
+
+static const unsigned char ttable[6][4] = {
+  // R_START (00)
+  {R_START_M,            R_CW_BEGIN,     R_CCW_BEGIN,  R_START},
+  // R_CCW_BEGIN
+  {R_START_M | DIR_CCW, R_START,        R_CCW_BEGIN,  R_START},
+  // R_CW_BEGIN
+  {R_START_M | DIR_CW,  R_CW_BEGIN,     R_START,      R_START},
+  // R_START_M (11)
+  {R_START_M,            R_CCW_BEGIN_M,  R_CW_BEGIN_M, R_START},
+  // R_CW_BEGIN_M
+  {R_START_M,            R_START_M,      R_CW_BEGIN_M, R_START | DIR_CW},
+  // R_CCW_BEGIN_M
+  {R_START_M,            R_CCW_BEGIN_M,  R_START_M,    R_START | DIR_CCW},
+};
+#else
+// Use the full-step state table (emits a code at 00 only)
+#define R_CW_FINAL 0x1
+#define R_CW_BEGIN 0x2
+#define R_CW_NEXT 0x3
+#define R_CCW_BEGIN 0x4
+#define R_CCW_FINAL 0x5
+#define R_CCW_NEXT 0x6
+
+static const unsigned char ttable[7][4] = {
+  // R_START
+  {R_START,    R_CW_BEGIN,  R_CCW_BEGIN, R_START},
+  // R_CW_FINAL
+  {R_CW_NEXT,  R_START,     R_CW_FINAL,  R_START | DIR_CW},
+  // R_CW_BEGIN
+  {R_CW_NEXT,  R_CW_BEGIN,  R_START,     R_START},
+  // R_CW_NEXT
+  {R_CW_NEXT,  R_CW_BEGIN,  R_CW_FINAL,  R_START},
+  // R_CCW_BEGIN
+  {R_CCW_NEXT, R_START,     R_CCW_BEGIN, R_START},
+  // R_CCW_FINAL
+  {R_CCW_NEXT, R_CCW_FINAL, R_START,     R_START | DIR_CCW},
+  // R_CCW_NEXT
+  {R_CCW_NEXT, R_CCW_FINAL, R_CCW_BEGIN, R_START},
+};
+#endif
+
+static void rotary_cb(void *rot_dev) {
+  rotary_t *dev = (rotary_t*) rot_dev;
+  unsigned int b1_v = gpio_read(dev->pin1) ? 1 : 0;
+  unsigned int b2_v = gpio_read(dev->pin2) ? 1 : 0;
+  unsigned char pinstate = ( b1_v? 2 : 0) | (b2_v ? 1 : 0);
```

Comment:
`b1_v` and `b2_v` don't need to be set to 0 or 1. Simply use
```C
int b1_v = gpio_read(dev->pin1);
int b2_v = gpio_read(dev->pin2);
```

or get rid of them :-)

Additionally, I would also make `pinstate` an `int` since it's only a support variable. If you remove the other 2 variables there's no size impact (and sometimes avoids extra casting introduced by the compiler).

```C
int pinstate = (gpio_read(dev->pin1) ? 2 : 0) | (gpio_read(dev->pin2) ? 1 : 0);
```